driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -997,4 +1095,69 @@ private long awaitNanos(final long timeoutNanos) {
         private MutableReference() {
         }
     }
+
+    private ServerStats getServerStats(final ObjectId serviceId) {
+        ServerStats serverStats = serverStatsMap.get(serviceId);
+        if (serverStats == null) {
+            serverStats = new ServerStats();
+            serverStatsMap.put(serviceId, serverStats);
+        }
+        return serverStats;
+    }
+
+    private synchronized void addConnectionToServerStats(final ObjectId serviceId) {
```

Comment:
We have reads and writes related to different services all contending on the same exclusive lock. It would be nice to avoid this by using `ConcurrentHashMap`, but I can't see immediately how to achieve that in this case without starting to produce garbage `ServerStats` objects on each write. I will think more about this, and maybe someone else would also be interested to consider whether this is doable without too much efforts (btw, we may decide that producing garbage `ServerStats` objects may be tolerable provided how much garbage is produced while a connection is used between checkout and checkin).

Reply:
Turned out, that `ConcurrentHashMap` is not sufficiently documented, and the documentation was improved in the Java SE 14 (see [JDK-8231592 Clarify that ConcurrentHashMap compute methods mapping functions execute at most once](https://bugs.openjdk.java.net/browse/JDK-8231592)). Thus, we can rely on the guarantee that the [_"supplied function is invoked exactly once per invocation of this method"_](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#compute(K,java.util.function.BiFunction)) specified in the Java SE 14 even though we must be able to run on an implementation of the Java SE 8 (this guarantee is not a result of the change of the CMH behavior, but merely a clarification of the behavior as it was since the method was added). With this in mind, we can modify `ServerStats` in the function passed to `CHM.compute` instead of creating a new instance of `ServerStats` (as I was imagining previously, and as we would have to do if we were relying only on the guarantees of the `ConcurrentMap` interface).

Thus, I propose the following:
```java
@ThreadSafe
private static class ServiceStateManager {
    private final ConcurrentHashMap<ObjectId, ServiceState> stateByServiceId = new ConcurrentHashMap<>();

    void addConnection(ObjectId serviceId) {
        stateByServiceId.compute(serviceId, (k, v) -> {
            if (v == null) {
                v = new ServiceState();
            }
            v.incrementConnectionCount();
            return v;
        });
    }

    /**
     * Removes the mapping from {@code serviceId} to a {@link ServiceState} if its connection count reaches 0.
     * This is done to prevent memory leaks.
     * <p>
     * This method must be called once for any connection for which {@link #addConnection(ObjectId)} was called.
     */
    void removeConnection(ObjectId serviceId) {
        stateByServiceId.compute(serviceId, (k, v) -> {
            assertNotNull(v);
            return v.decrementAndGetConnectionCount() == 0 ? null : v;
        });
    }

    /**
     * This method must not be called after calling {@link #removeConnection(ObjectId)} for any connection.
     * With this restriction, we still may use {@code expectedGeneration} to avoid incrementing the generation
     * if {@code expectedGeneration} is outdated, despite the generation not not growing monotonically for any given service
     * due to method {@link #removeConnection(ObjectId)} removing {@link ServiceState} when the connection count reaches 0.
     */
    boolean incrementGeneration(ObjectId serviceId, int expectedGeneration) {
        ServiceState state = stateByServiceId.get(serviceId);
        assertNotNull(state);
        return state.incrementGeneration(expectedGeneration);
    }

    int getGeneration(ObjectId serviceId) {
        ServiceState state = stateByServiceId.get(serviceId);
        return state == null ? 0 : state.getGeneration();
    }

    @ThreadSafe
    private static class ServiceState {
        private final AtomicInteger generation = new AtomicInteger();
        private final AtomicInteger connectionCount = new AtomicInteger();

        void incrementConnectionCount() {
            connectionCount.incrementAndGet();
        }

        int decrementAndGetConnectionCount() {
            return connectionCount.decrementAndGet();
        }

        boolean incrementGeneration(int expectedGeneration) {
            return generation.compareAndSet(expectedGeneration, expectedGeneration + 1);
        }

        public int getGeneration() {
            return generation.get();
        }
    }
}
```

Reply:
Done

Reply:
What is the reason to have synchronization on the `DefaultConnectionPool` instance when calling `ServiceStateManager.getGeneration`/`addConnection/,`removeConnection` (worth noting that there is currently no such synchronization when calling `ServiceStateManager.incrementGeneration`). I think none of those are needed, and actually they annihilate the attempt to avoid users of connections to different services from contending for the same lock.

Reply:
It was an incorrect refactoring.  Fixed in subsequent commit

Reply:
The current implementation of `ServiceStateManager.incrementGeneration`
```java
boolean incrementGeneration(final ObjectId serviceId, final int expectedGeneration) {
    ServiceState state = stateByServiceId.compute(serviceId, (k, v) -> {
        if (v == null) {
            v = new ServiceState();
        }
        return v;
    });
    return state.incrementGeneration(expectedGeneration);
}
```
is susceptible to memory leaks. The assertion `assertNotNull(v)` was removed because it does not hold when `incrementGeneration` is called from `closeAndHandleOpenFailure`. However, if we store `new ServiceState()` for such situations, we may end up in a situation when this mapping is never removed because the service went away and there were no more `removeConnection` calls for this `serviceId`. I propose to use the following implementation, which cannot create a memory leak while still tolerating absence of a mapping for `serviceId`:
```java
boolean incrementGeneration(final ObjectId serviceId, final int expectedGeneration) {
    ServiceState state = stateByServiceId.get(serviceId);
    return state != null && state.incrementGeneration(expectedGeneration);
}
```